tests/tests/test_state_scripts.py

```diff
@@ -1170,17 +1137,21 @@ def verify_script_log_correct(self, test_set, log):
                 if (log_pos < len(log)
                     and log[log_pos - 1].startswith("Sync_")
                     and log[log_pos].startswith("Idle_")
-                    and not expected_order[expected_pos].startswith("Idle_")):
+                    and not expected_order[expected_pos % len(expected_order)].startswith("Idle_")):
                     # The Idle/Sync sequence is allowed to "wrap around" and start
                     # over, because it may take a few rounds of checking before the
                     # deployment is ready for the client.
-                    expected_pos = idle_pos
+                    expected_pos = 0
+
+                if expected_pos == len(expected_order):
+                    num_iterations = num_iterations + 1
+                    expected_pos = 0
```

Comment:
I have to admit that I have some trouble following: You now set `expected_pos` back to zero, which will start from the beginning of the test_set? But won't that be wrong? We don't want the entire sequence to repeat, at least not the ones that get beyond `Download_Enter`. Can you explain a bit more?

Reply:
Right, I will then document the code inline.

The point here is that I want to measure the total number of iterations that the `log` list wraps around `expected_order` list (also taking into account the Idle/Sync wrapping allowed inside each iteration).

The previous code will only ensured that `log` list included `expected_order`, but didn't care about how many times.

Reply:
> We don't want the entire sequence to repeat, at least not the ones that get beyond `Download_Enter`. Can you explain a bit more?

Exactly, we don't want it to repeat. Hence we save the number of iterations (repetitions!) of the entire sequence and then verify that it has repeated only _few_ times.

Then, the "few" concept is quite complicated because sometimes the test frameworks takes very long to stop the client, so this can go several iterations of 5s each through the sequence. Hence my upper limit of 20 iterations below.

Does it make sense?

Reply:
I understand what you mean, but all the states starting with `Artifact` should execute exactly once, and never more than once, not even when the tests are slow. That's why I'm questioning whether setting `expected_pos` is correct, because we should not allow any repeated sequences of those states.

Do you agree that it is only the `Idle` / `Sync` sequence that should repeat?

Reply:
Hi again,

So I reworked the test following our conversation couple of days ago.

Working on the code, I figured out that I didn't need a new parameter in the test sets: using `ExpectedStatus` will actually say if the test is expect to loop (i.e. expected None) or not (i.e. expected success/failure). The test logic uses now this parameter to strictly verify when the case is allowed to loop or not.

There is an additional change I need you to verify. I think that the status expectation for `Failure_in_Download_Leave_script` was wrong, as it should be "failure". Please confirm or argue back :)

Reply:
> I think that the status expectation for `Failure_in_Download_Leave_script` was wrong, as it should be "failure". Please confirm or argue back

You're absolutely right! I wondered why it passed in the first place, but I guess it wasn't checked at all if it was `None`.

As for the rest, yes, I believe you nailed it now. Clever to use `ExpectedStatus` as an indication for number of iterations!

Reply:
Thanks! Good that we understood each other at the end :)